src/core/EventDispatcher.js

```diff
@@ -6,79 +6,52 @@ class EventDispatcher {
 
 	addEventListener( type, listener ) {
 
-		if ( this._listeners === undefined ) this._listeners = {};
+		const listeners = this.getEventListeners( type );
 
-		const listeners = this._listeners;
-
-		if ( listeners[ type ] === undefined ) {
-
-			listeners[ type ] = [];
-
-		}
-
-		if ( listeners[ type ].indexOf( listener ) === - 1 ) {
-
-			listeners[ type ].push( listener );
-
-		}
+		listeners.add( listener );
 
 	}
 
 	hasEventListener( type, listener ) {
 
-		if ( this._listeners === undefined ) return false;
-
-		const listeners = this._listeners;
-
-		return listeners[ type ] !== undefined && listeners[ type ].indexOf( listener ) !== - 1;
+		return this.getEventListeners( type ).has( listener );
 
 	}
 
 	removeEventListener( type, listener ) {
 
-		if ( this._listeners === undefined ) return;
-
-		const listeners = this._listeners;
-		const listenerArray = listeners[ type ];
+		this.getEventListeners( type ).remove( listener );
 
-		if ( listenerArray !== undefined ) {
+	}
 
-			const index = listenerArray.indexOf( listener );
+	dispatchEvent( event ) {
 
-			if ( index !== - 1 ) {
+		event.target = this;
 
-				listenerArray.splice( index, 1 );
+		const invokeListener = listener => listener.call( this, event );
```

Comment:
Would it be possible to avoid the definition of an inline function each time `dispatchEvent()` is executed.

Besides, the listener array was originally cloned to avoid breakage when listeners are removed during the iteration. It seems this fix is missing in the new approach, doesn't it?